### PR TITLE
Broaden UIGraphicsRenderer Experiment to Everywhere

### PR DESCRIPTION
**Pull request by Adlai-Holler**
This gets rid of the ASExperimentalGraphicsContext experiment that was a dud, and replaces ASGraphicsContexts.h with a wrapper. The wrapper calls into either UIGraphicsBeginImageContext or UIGraphicsImageRenderer to build an image.

These graphics renderers seem pretty awesome. https://gist.github.com/NSProgrammer/38f1b4f9b29c3077af3f39a7e98092b1

Want multiple core-team reviews on this since it touches such core code. It's pretty straightforward though. The one case where it wasn't straightforward (`__didDisplay…`) I just left with UIGraphics.

### CODE REVIEW

**Comment by Adlai-Holler** [comment]
This diff is a pain to review, sorry. Split view makes it a lot easier. One day github's diff will learn to ignore whitespace – it's only Microsoft, after all.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/be57015e02e076ed17278fd56b2edaeb/raw/e8504ff55bc492f8a3d7628f32af71000bac0125/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/4e09eb640d6d83b8e01cf0a3b63d9d85/raw/f6ec7b1f464e24ea47815ae05e98f96d5b7a737f/log.txt)

**Comment by wiseoldduck** [comment]
> This diff is a pain to review, sorry. Split view makes it a lot easier. One day github's diff will learn to ignore whitespace – it's only Microsoft, after all.
Screen Shot 2019-03-29 at 4 53 44 PM
FWIW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/badc6309350f061f6dbabbe06577f621/raw/ad980bbacb73e5806316a1a2ec44e5911fc1a98b/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
There's that old-duck wisdom

File: Source/ASImageNode.mm [hunk 1]
```diff
@@ -483,53 +480,45 @@ + (UIImage *)createContentsForkey:(ASImageNodeContentsKey *)key drawParameters:(
...
-  CGBlendMode blendMode = canUseCopy ? kCGBlendModeCopy : kCGBlendModeNormal;
-  
-  @synchronized(image) {
-    [image drawInRect:imageDrawRect blendMode:blendMode alpha:1];
-  }
-  
-  if (context && key.didDisplayNodeContentWithRenderingContext) {
-    key.didDisplayNodeContentWithRenderingContext(context, drawParameters);
-  }
+  UIImage *result = ASGraphicsCreateImageWithOptions(key.backingSize, key.isOpaque, 1.0, image, ^{
```
Comment (wiseoldduck) [inline Source/ASImageNode.mm]: This previously would not choose opaque if the image wouldn't fill the context (now a new UIImage, but apparently sized the same). Is this change intentional?

**Comment by wiseoldduck** [comment]
`ASExperimentalGraphicsContexts` is gone :shrug: Did we determine it was not helpful, or are we just sure this new experiment is good enough we won't care about that anymore?

LGTM overall

File: Source/ASImageNode.mm (hunk 1 above)
Comment (Adlai-Holler) [reply Source/ASImageNode.mm]: Oh good call, this was a bad merge. Fixed

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
> `ASExperimentalGraphicsContexts` is gone :shrug: Did we determine it was not helpful, or are we just sure this new experiment is good enough we won't care about that anymore?
> 
> LGTM overall

It wasn't helpful in the first place. Could be that the options were wrong on some platform, or whatever, but everyone turned it off.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASExperimentalFeatures.h [hunk 2]
```diff
@@ -16,24 +16,23 @@ NS_ASSUME_NONNULL_BEGIN
...
+  ASExperimentalFramesetterCache = 1 << 5,                  // exp_framesetter_cache
+  ASExperimentalSkipClearData = 1 << 6,                     // exp_skip_clear_data
+  ASExperimentalDidEnterPreloadSkipASMLayout = 1 << 7,      // exp_did_enter_preload_skip_asm_layout
+  ASExperimentalDisableAccessibilityCache = 1 << 8,         // exp_disable_a11y_cache
+  ASExperimentalDispatchApply = 1 << 9,                     // exp_dispatch_apply
+  ASExperimentalImageDownloaderPriority = 1 << 10,          // exp_image_downloader_priority
+  ASExperimentalDrawing = 1 << 11,                          // exp_drawing
+  ASExperimentalFixRangeController = 1 << 12,               // exp_fix_range_controller
+  ASExperimentalOOMBackgroundDeallocDisable = 1 << 13,      // exp_oom_bg_dealloc_disable
+  ASExperimentalTransactionOperationRetainCycle = 1 << 14,  // exp_transaction_operation_retain_cycle
```
Comment (bolsinga) [inline Source/ASExperimentalFeatures.h]: Should re-ordering these be done with a separate diff?

File: Source/ASTextNode.mm [hunk 3]
```diff
@@ -538,64 +538,30 @@ + (UIImage *)displayWithParameters:(id<NSObject>)parameters isCancelled:(NS_NOES
 {
   ASTextNodeDrawParameter *drawParameter = (ASTextNodeDrawParameter *)parameters;
   
-  if (drawParameter->_bounds.size.width <= 0 || drawParameter->_bounds.size.height <= 0) {
+  if (CGRectIsEmpty(drawParameter->_bounds)) {
```
Comment (bolsinga) [inline Source/ASTextNode.mm]: Should this be done with a separate diff?

File: Source/ASTextNode.mm [hunk 4]
```diff
@@ -538,64 +538,30 @@ + (UIImage *)displayWithParameters:(id<NSObject>)parameters isCancelled:(NS_NOES
...
-        [renderer drawInContext:context bounds:drawParameter->_bounds];
-        CGContextRestoreGState(context);
-      }];
-    }
-  }
-  
-  if (!renderedWithGraphicsRenderer) {
-    UIGraphicsBeginImageContextWithOptions(CGSizeMake(drawParameter->_bounds.size.width, drawParameter->_bounds.size.height), drawParameter->_opaque, drawParameter->_contentScale);
-      
+  return ASGraphicsCreateImageWithOptions(drawParameter->_bounds.size, drawParameter->_opaque, drawParameter->_contentScale, nil, ^{
```
Comment (bolsinga) [inline Source/ASTextNode.mm]: @garrettmoon is still in the midst of rolling this experiment out. Please let him review before landing if possible.

File: Source/ASTextNode.mm (hunk 3 above)
Comment (Adlai-Holler) [reply Source/ASTextNode.mm]: I don't know. Should it?

File: Source/ASTextNode.mm (hunk 4 above)
Comment (Adlai-Holler) [reply Source/ASTextNode.mm]: Yep, won't land without Garrett's take. I can break this apart into "TextDrawing" and "NonTextDrawing" if that is useful.

**Comment by Adlai-Holler** [comment]
Note: don't land until @garrettmoon weighs in. Garrett, it's not a problem to leave your experiment and change this diff to add a separate one for the rest of the drawing. Let me know which way you want to go.

File: Source/ASExperimentalFeatures.h (hunk 2 above)
Comment (Adlai-Holler) [reply Source/ASExperimentalFeatures.h]: This is not re-ordering, it's replacing and it's an essential part of this diff.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/76c8d2b1c833bbc701939072b3a43fe2/raw/d924eefcd8b884919cd29069f70c72293cd2e067/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/0e017d1dca480bb726679b3e8fcc50e8/raw/e36a0d4dbee9bf59a32aa42ca7c4b4bdf3f232e8/log.txt)

**Comment by garrettmoon** [comment]
I'd like to get a bit more data from the text experiment before we merge this. Thank you @Adlai-Holler for putting this up!

**Comment by Adlai-Holler** [comment]
OK let's leave your experiment intact. Reopening as a separate experiment.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/6f32a629bfaba3cb4b38650857548c78/raw/2fc57f9a35a8840c574950240001fbb17411d281/log.txt)
